### PR TITLE
Make kernel hash stable in type reduced build

### PR DESCRIPTION
**Pull request by skottmckay**
**Description**: 
Add infrastructure so that a kernel definition has the full list of supported types and a list of types enabled in this build. We need to use the full list when calculating the kernel hash so that the hash value in an ORT format model is stable across builds with and without type reduction enabled.

Update the kernels that support type reduction within the kernel implementation to provide both lists.

**Motivation and Context**
ORT format model created with a full build (which is the typical usage) would have a different hash for a kernel than a type reduced build due to the type constraints changing. This would prevent the model from being able to be loaded.

### CODE REVIEW

**Review by yuslepukhin — DISMISSED** [review]
:shipit:

File: onnxruntime/core/providers/op_kernel_type_control.h [hunk 1]
```diff
@@ -186,6 +186,40 @@ struct EnabledTypes {
...
+ * @param ArgIndex Index of the given Op kernel argument.
+ */
+#define ORT_OP_KERNEL_ARG_SUPPORTED_TYPE_LIST(                                                          \
+    OpProvider, OpDomain, OpName, OpSet, ArgDirection, ArgIndex)                                        \
+  ::onnxruntime::op_kernel_type_control::TypesHolder<                                                   \
+      ::onnxruntime::op_kernel_type_control::tags::Supported<                                           \
+          ORT_OP_KERNEL_TYPE_CTRL_INTERNAL_OP_KERNEL_ARG_TAG(OpDomain, OpName, ArgDirection, ArgIndex), \
+          ::onnxruntime::op_kernel_type_control::                                                       \
+              ORT_OP_KERNEL_TYPE_CTRL_INTERNAL_PROVIDER_TAG_CLASS_NAME(OpProvider),                     \
+          OpSet>>::types
```
Comment (edgchen1) [inline onnxruntime/core/providers/op_kernel_type_control.h:206]: could we put the Supported TypesHolder type into a separate macro (perhaps ORT_OP_KERNEL_TYPE_CTRL_INTERNAL_SUPPORTED_TYPES_HOLDER) and use that here and in ORT_OP_KERNEL_ARG_ENABLED_TYPE_LIST?

File: include/onnxruntime/core/framework/kernel_def_builder.h [hunk 2]
```diff
@@ -128,7 +128,12 @@ class KernelDef {
   // The supported data types for inputs/outputs.
   // Key is input/output name defined in op schema, Value are supported types.
   // note: std::map as we need the order to be deterministic for the hash
-  std::map<std::string, std::vector<MLDataType>> type_constraints_;
+  // Note: supported_type_constraints_ are used to calculate the kernel hash so that the hash is
```
Comment (edgchen1) [inline include/onnxruntime/core/framework/kernel_def_builder.h:131]: >  [](start = 96, length = 1)

nit: trailing space

**Review by edgchen1 — APPROVED** [review]
:shipit:

**Pull request merged**